Allow offline bootstrapping of cabal-install

sheaf opened the pull request:
This ports to cabal-install the offline bootstrapping logic which was
introduced for Hadrian in GHC MR !6315.

This adds a "fetch" command to the bootstrap script, which fetches
all the dependency tarballs from Hackage, to be used in an offline
build. See bootstrap/README.md for further information.

I tested this locally but would appreciate further testing, perhaps changing the bootstrap CI scripts to separately fetch the archive and then build with the archives.

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

wz1000 commented on bootstrap/bootstrap.py:
```diff
@@ -79,10 +89,10 @@ def __init__(self, ghc_path: Path):
         #self.lib_dir = Path(info['LibDir'])
         #self.ghc_pkg_path = (self.lib_dir / 'bin' / 'ghc-pkg').resolve()
-        self.ghc_pkg_path = (self.ghc_path.parent / 'ghc-pkg').resolve()
+        self.ghc_pkg_path = (self.ghc_path.parent / 'ghc-pkg.exe').resolve()
```
The `.exe` suffix should probably be platform dependent?

sheaf replied:
oops I committed that by mistake, what's the usual way to do this in Python?

sheaf replied:
```python
        exe = ''
        if platform.system() == 'Windows': exe = '.exe'

        self.ghc_pkg_path = (self.ghc_path.parent / 'ghc-pkg' + exe).resolve()
```

?

sheaf replied:
That doesn't work, nor does `os.path.join(.., exe)`

sheaf replied:
OK done

wz1000 replied:
yeah, that seems fine.

Do we need the same change in the GHC bootstrap script to get it working on windows?

sheaf replied:
yes, that's in GHC MR !8821

Mikolaj commented:
Thanks a lot. Good point about "changing the bootstrap CI scripts to separately fetch the archive and then build with the archives". That would tick the "test" box required to accept the PR. For the "changelog" box, it's enough to include a tiny file describing the change, as in all other PRs.

robx commented on bootstrap/bootstrap.py:
```diff
@@ -334,55 +320,146 @@ def make_archive(cabal_path):
     return archivename
 
+def fetch_from_plan(plan : FetchPlan, output_dir : Path):
+  import urllib.request
```
Move this to the top level, start of the file? (I see it was inlined before, but I think it's not good to hide possible side effects of importing a module inside a function body.)

Also the indentation doesn't match the rest of the file

robx commented on bootstrap/bootstrap.py:
```diff
@@ -334,55 +320,146 @@ def make_archive(cabal_path):
+    return sources_dict
+
+def find_ghc(compiler) -> Compiler:
+  # Find compiler
```
this comment could be deleted

robx commented on bootstrap/bootstrap.py:
```diff
@@ -334,55 +320,146 @@ def make_archive(cabal_path):
-        ghc = Compiler(Path(path))
-    else:
-        ghc = Compiler(args.with_compiler)
+    subparsers = parser.add_subparsers(dest="command")
```
would it be an option to add a (default) `build` command? seems that would make the argument structure clearer

sheaf replied:
I can do this, but it would break existing users, so I would prefer not to.

robx replied:
I'm suggesting

```
$ ./bootstrap.py -h
usage: bootstrap.py [-h] [-d DEPS] [-w WITH_COMPILER] {fetch,build} ...

bootstrapping utility for cabal-install.

positional arguments:
  {fetch,build}
    fetch               fetch all required sources from Hackage (for offline builds)
    build               bootstrap build (default)

optional arguments:
  -h, --help            show this help message and exit
  -d DEPS, --deps DEPS  bootstrap dependency file
  -w WITH_COMPILER, --with-compiler WITH_COMPILER
                        path to GHC
```

The code would continue to do

```
if args.command == 'fetch':
    ...
else: # command == 'build' or it's not set
    ...
```

That wouldn't break existing users, would it?

sheaf replied:
OK, I've done that. Thanks.

robx commented on bootstrap/bootstrap.py:
```diff
@@ -334,55 +320,146 @@ def make_archive(cabal_path):
-    print(f'Bootstrapping cabal-install with GHC {ghc.version} at {ghc.ghc_path}...')
+    args = parser.parse_args()
 
     print(dedent("""
```
I don't have strong opinions on the use of `dedent`, but we should use it or not use it consistently. (I.e., suggesting to convert the multitline prints below to use `dedent` also, or possibly remove it everywhere.)

robx reviewed (commented):
I haven't tested it, but it seems generally reasonably.

Noted some minor suggestions.

Would it make sense to add coverage for this path to our CI testing of the bootstrap?

Mikolaj commented:
Let me rebase to include a workaround to our CI instability and fix the tests.

Mikolaj commented:
@mergify rebase

sheaf commented:
- Added a changelog entry (to `cabal-install-3.8.0.20220526.md`, let me know if it should be somewhere else).
- Made the bootstrap test use this new functionality.
- Addressed review comments.

fgaz commented:
> Added a changelog entry (to cabal-install-3.8.0.20220526.md, let me know if it should be somewhere else)

see https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog

robx commented on bootstrap/bootstrap.py:
```diff
@@ -334,55 +321,142 @@ def make_archive(cabal_path):
-    cabal_path = (BINDIR / 'cabal').resolve()
 
-    archive = make_archive(cabal_path)
+    if(args.command == 'fetch'):
```
```suggestion
    if args.command == 'fetch':
```

sheaf commented:
Updated the changelog, grazie Francesco.

robx reviewed (approved):
LGTM

sheaf commented:
Could we get this merged?

Mikolaj reviewed (approved):
Yes, let's merge. Thanks a lot!

Mikolaj commented:
I mean, after the conflict is resolved. :)
